LFS-320: Add support for notes on answers

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-320)

**Test branch**: `LFS-320-test` 

**To test**: From the homepage, create a Demographics form. The Family question has comments enabled, which can be added to.

**Additional notes**: This is configurable from the Questionnaire definition and is off by default. All question components that use an `Answer` component will have this (optionally) enabled, and for best-looking results I've moved the `Answer` component to the bottom of every question (it previously was sometimes at the top of the question, and sometimes at the bottom -- its positioning did not previously matter since it had no visible components)

OrdiNeu committed:
LFS-320: Add support for comments on answers
Add comments on any answer as long as enableComments is true for the
question.

sdumitriu reviewed (commented):
Didn't review yet, just a few notes:

- Do table filters take into account the comments? Probably not, we need to think how to handle that
- Does quick search match on comments?

marta- reviewed (changes requested):
Looks good so far!

Some UI improvements:
* [x] The button's label should always be "Notes" (no show/hide, and we'll call them notes instead of comments)
* [x] The text should not be all uppercase (please override the default button style)
* [x] The icon should be:
 - "+" if the comment is empty and collapsed
 - the current "expand" icon if the comment is NOT empty and collapsed
 - the current "collapse" icon if the coment is expanded (empty or not)
* [x] The button needs a tooltip: "Add notes" / "Show notes" / "Hide notes"

After #168 is merged, please rebase and make sure the comment match is displayed properly.

marta- commented:
> * Do table filters take into account the comments? Probably not, we need to think how to handle that
Tested, they do, but we should discuss making it more clear to the user

> * Does quick search match on comments?
Tested, it does, but the match probably won't show properly (to be confirmed after merging #168 and rebasing this PR.

OrdiNeu committed:
LFS-320: Add support for comments on answers
Restyle the show comment button, and allow it to store no data as well.

OrdiNeu committed:
LFS-320: Add support for comments on answers
Delete comments if nothing is entered.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Comment.jsx:
```diff
@@ -0,0 +1,85 @@
+        >
+        <Button
+          color = "default"
+          className = {classes.showCommentsButton}
```
```suggestion
          className = {classes.toggleCommentsButton}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Comment.jsx:
```diff
@@ -0,0 +1,85 @@
+  const commentIsEmpty = comment == null || comment == "";
+
+  return (<React.Fragment>
+    <div className = {classes.showCommentsContainer}>
```
```suggestion
    <div className = {classes.commentsContainer}>
```

OrdiNeu replied:
This `<div>` only contains the toggle notes button and not the comments themselves -- maybe I'll try `toggleCommentsContainer`?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -169,6 +169,16 @@ const questionnaireStyle = theme => ({
     highlightedTitle: {
         color: theme.palette.primary.light,
+    },
+    showCommentsContainer: {
```
```suggestion
    commentsContainer: {
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -169,6 +169,16 @@ const questionnaireStyle = theme => ({
+    showCommentsContainer: {
+        padding: theme.spacing(1, 0, 1, 2)
+    },
+    showCommentsButton: {
```
```suggestion
    toggleCommentsButton: {
```

marta- commented:
@sashaandjic @OrdiNeu @k-pace  What is your opinion about "Comments" vs "Notes" for labeling this text field? It started out as "Comments", but I switched it to "Notes" because it seems to make more sense now. Your input is appreciated.

![captured (6)](https://user-images.githubusercontent.com/651980/75926109-ef58ae00-5e37-11ea-85e2-04f05267feed.gif)

sashaandjic commented:
> @sashaandjic @OrdiNeu @k-pace What is your opinion about "Comments" vs "Notes" for labeling this text field? It started out as "Comments", but I switched it to "Notes" because it seems to make more sense now. Your input is appreciated.
> 
> ![captured (6)](https://user-images.githubusercontent.com/651980/75926109-ef58ae00-5e37-11ea-85e2-04f05267feed.gif)

My vote goes to Notes

marta- commented:
I kind of like the clean look of the + icon, and it's consistent with the expand/collapse. I see what you mean about consistency with other (+) buttons, however those are more "global" actions. I would leave this one as is and maybe consider simplifying the filters + button instead.

sashaandjic commented:
How about movin the (+) and colapse bttns to the right?
![image](https://user-images.githubusercontent.com/23243244/75928413-59735200-5e3c-11ea-8d8f-e4a12297f18b.png)

k-pace commented:
1. I like 'Notes' best
2. I like the check on the left side best

marta- commented:
Okay, we'll keep it  looking like it does now. @OrdiNeu please rename all variables / components / cas classes to turn 'comments' into 'notes'. Thanks!

OrdiNeu committed:
LFS-320: Add support for comments on answers

Apply suggestions from @marta-

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

OrdiNeu committed:
LFS-320: Add support for comments on answers
Rename "Comments" to "Notes" globally

OrdiNeu commented:
> Okay, we'll keep it looking like it does now. @OrdiNeu please rename all variables / components / css classes to turn 'comments' into 'notes'. Thanks!

Changes are complete, but I've switched `showCommentsContainer` to `toggleNotesContainer` instead of dropping the prefix altogether, since it only affects the button

sashaandjic commented:
I'd like to see when the focus is on the Notes button
![image](https://user-images.githubusercontent.com/23243244/76102428-ff8a9d80-5f9d-11ea-86f9-d2dae8333e56.png)

marta- commented:
> > * Does quick search match on comments?
> 
> Tested, it does, but the match probably won't show properly (to be confirmed after merging #168 and rebasing this PR).

This is no longer true after rebase. @sdumitriu wdyt, should we handle this now or as part of the indexing fixes?

marta- commented:
> I'd like to see when the focus is on the Notes button

@sashaandjic I don't follow. This is how it behaves for me (highlight on hover and click):

![captured (1)](https://user-images.githubusercontent.com/651980/76105901-d967fc00-5fa3-11ea-9be8-d48963c41064.gif)

sashaandjic commented:
@marta- when tabbing from Family filed, you will end up with the situation presented in the screeshoot

marta- commented:
However it would be nice to autofocus on the textarea when opening the notes.

marta- commented:
> @marta- when tabbing from Family filed, you will end up with the situation presented in the screeshoot

@sashaandjic I see. This looks like a Material UI limitation, I wouldn't get into fixing it right now. But please create a task so that we can tackle it globally when we have time.

marta- commented:
@OrdiNeu with the decision that the quick search issue would be handled in a separate PR, the approval on my side is pending this minor usability improvement:
> However it would be nice to autofocus on the textarea when opening the notes.

OrdiNeu committed:
LFS-320: Add support for comments on answers
Autofocus on the textarea when opening the notes.
Also re-enable the focus ripple.

OrdiNeu commented:
>I'd like to see when the focus is on the Notes button

> However it would be nice to autofocus on the textarea when opening the notes.

Both of these should be fixed in ` eb30f721`.

sashaandjic reviewed (approved):
tested

marta- reviewed (approved):
Looks great
